lib/service/signals.go

```diff
@@ -0,0 +1,338 @@
+	l, err := process.importListener(listenerType, address)
+	if err == nil {
+		log.Infof("Using file descriptor %v %v passed by the parent process.", listenerType, address)
+		return l, nil
+	}
+	if !trace.IsNotFound(err) {
+		return nil, trace.Wrap(err)
+	}
+	log.Infof("Service %v is creating new listener on %v.", listenerType, address)
+	return process.createListener(listenerType, address)
+}
+
+// importListener imports listener passed by the parent process, if no listener is found
+// returns NotFound, otherwise removes the file from the list
+func (process *TeleportProcess) importListener(listenerType, address string) (net.Listener, error) {
+	process.Lock()
+	defer process.Unlock()
+
+	for i := range process.importedDescriptors {
+		d := process.importedDescriptors[i]
+		if d.Type == listenerType && d.Address == address {
+			l, err := d.ToListener()
+			if err != nil {
+				return nil, trace.Wrap(err)
+			}
+			process.importedDescriptors = append(process.importedDescriptors[:i], process.importedDescriptors[i+1:]...)
+			process.registeredListeners = append(process.registeredListeners, RegisteredListener{Type: listenerType, Address: address, Listener: l})
+			return l, nil
+		}
+	}
+
+	return nil, trace.NotFound("no file descriptor for type %v and address %v has been imported", listenerType, address)
+}
+
+// createListener creates listener and adds to a list of tracked listeners
+func (process *TeleportProcess) createListener(listenerType, address string) (net.Listener, error) {
+	listener, err := net.Listen("tcp", address)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	process.Lock()
+	defer process.Unlock()
+	r := RegisteredListener{Type: listenerType, Address: address, Listener: listener}
+	process.registeredListeners = append(process.registeredListeners, r)
+	return listener, nil
+}
+
+// exportFileDescriptors exports file descriptors to be passed to child process
+func (process *TeleportProcess) exportFileDescriptors() ([]FileDescriptor, error) {
+	var out []FileDescriptor
+	process.Lock()
+	defer process.Unlock()
+	for _, r := range process.registeredListeners {
+		file, err := utils.GetListenerFile(r.Listener)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+		out = append(out, FileDescriptor{File: file, Type: r.Type, Address: r.Address})
+	}
+	return out, nil
+}
+
+// importFileDescriptors imports file descriptors from environment if there are any
+func importFileDescriptors() ([]FileDescriptor, error) {
+	// These files may be passed in by the parent process
+	filesString := os.Getenv(teleportFilesEnvVar)
+	if filesString == "" {
+		return nil, nil
+	}
+
+	files, err := filesFromString(filesString)
+	if err != nil {
+		return nil, trace.BadParameter("child process has failed to read files, error %q", err)
+	}
+
+	if len(files) != 0 {
+		log.Infof("Child has been passed files: %v", files)
+	}
+
+	return files, nil
+}
+
+type RegisteredListener struct {
```

Comment:
Can you add godocs on exported structs and fields? (`RegisteredListener` and `FileDescriptor` below.)